### PR TITLE
Support DynamicSlicing in torchdynamo

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90410

Currently, i am limiting the step size to be normal integer, just because I am not really sure what the expected behaviour for x[None:None:symint] as the original python behaviour adjusts start and stop based on step. 

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

Differential Revision: [D42049013](https://our.internmc.facebook.com/intern/diff/D42049013)

### CODE REVIEW

File: torch/csrc/autograd/python_variable_indexing.h [hunk 1]
```diff
@@ -1,11 +1,91 @@
...
 namespace autograd {
 
+struct UnpackedSlice {
+  c10::SymInt start;
+  c10::SymInt stop;
+  c10::SymInt step;
+};
+
+// This mirrors Cpython's PySlice_Unpack method
+inline UnpackedSlice __PySlice_Unpack(PyObject* _r) {
```
Comment (zhxchen17) [inline torch/csrc/autograd/python_variable_indexing.h:21]: Is inlining this function intentional? I think it makes more sense to put definition into cpp file.

**Comment by voznesenskym** [comment]
Lots of test failures :(

File: torch/csrc/autograd/python_variable_indexing.h [hunk 2]
```diff
@@ -1,11 +1,91 @@
...
+  PySliceObject* r = (PySliceObject*)_r;
```
Comment (voznesenskym) [inline torch/csrc/autograd/python_variable_indexing.h:22]: Why did this change files? This makes it harder to review...

**Comment by voznesenskym** [comment]
If this needs ATen / core native changes, change that first + add tests + try to minimize code motion there. Then do the dynamo change on top of that.

**Review by voznesenskym — CHANGES_REQUESTED** [review]
If this needs ATen / core native changes, change that first + add tests + try to minimize code motion there. Then do the dynamo change on top of that.

File: aten/src/ATen/TensorIndexing.h [hunk 3]
```diff
@@ -39,50 +39,65 @@ struct TORCH_API Slice final {
...
+        // guards against later undefined behaviour resulting from code that
+        // does "step = -step" as part of a slice reversal.
+        if (step_.expect_int() < -INDEX_MAX) {
+          step_ = c10::SymInt(-INDEX_MAX);
+        }
+      }
     }
     if (!start_index.has_value()) {
-      start_ = step_ < 0 ? INDEX_MAX : 0;
+      TORCH_CHECK_VALUE(!step_.is_symbolic(), "slice step cannot be symbolic");
```
Comment (albanD) [inline aten/src/ATen/TensorIndexing.h:61]: If you expect this here, then you don't need to handle the case above?

File: aten/src/ATen/TensorIndexing.h [hunk 4]
```diff
@@ -213,14 +228,14 @@ namespace impl {
...
-    int64_t step,
+    c10::SymInt start,
+    c10::SymInt stop,
+    c10::SymInt step,
     bool disable_slice_optimization,
     const at::Device& self_device,
     const c10::optional<SymIntArrayRef>& self_sizes) {
   // TODO: implement negative step
-  TORCH_CHECK_VALUE(step > 0, "step must be greater than zero");
+  TORCH_CHECK_VALUE(step.expect_int() > 0, "step must be greater than zero");
```
Comment (albanD) [inline aten/src/ATen/TensorIndexing.h:238]: No need for expect here either.

File: torch/csrc/autograd/python_variable_indexing.cpp [hunk 5]
```diff
@@ -133,14 +133,8 @@ inline Variable valueToTensor(
...
 
-static inline void checkUnpackSlice(
-    PyObject* index,
-    Py_ssize_t* start_ptr,
-    Py_ssize_t* stop_ptr,
-    Py_ssize_t* step_ptr) {
-  if (!THPUtils_unpackSlice(index, start_ptr, stop_ptr, step_ptr)) {
-    throw python_error();
-  }
+static inline UnpackedSlice checkUnpackSlice(PyObject* index) {
```
Comment (albanD) [inline torch/csrc/autograd/python_variable_indexing.cpp:136]: nit: this indirection is not needed anymore.

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42049013).

File: aten/src/ATen/TensorIndexing.h [hunk 6]
```diff
@@ -39,50 +39,65 @@ struct TORCH_API Slice final {
...
-      TORCH_CHECK_VALUE(step_ != 0, "slice step cannot be zero");
-
-      // Here step might be -INDEX_MAX-1; in this case we replace it
-      // with -INDEX_MAX.  This doesn't affect the semantics, and it
-      // guards against later undefined behaviour resulting from code that
-      // does "step = -step" as part of a slice reversal.
-      if (step_ < -INDEX_MAX)
-        step_ = -INDEX_MAX;
+      if (!step_.is_symbolic()) {
+        TORCH_CHECK_VALUE(step_.expect_int() != 0, "slice step cannot be zero");
```
[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply aten/src/ATen/TensorIndexing.h]: updated

File: aten/src/ATen/TensorIndexing.h [hunk 7]
```diff
@@ -39,50 +39,65 @@ struct TORCH_API Slice final {
...
       step_ = step_index.value();
-      TORCH_CHECK_VALUE(step_ != 0, "slice step cannot be zero");
-
-      // Here step might be -INDEX_MAX-1; in this case we replace it
-      // with -INDEX_MAX.  This doesn't affect the semantics, and it
-      // guards against later undefined behaviour resulting from code that
-      // does "step = -step" as part of a slice reversal.
-      if (step_ < -INDEX_MAX)
-        step_ = -INDEX_MAX;
+      if (!step_.is_symbolic()) {
```
[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply aten/src/ATen/TensorIndexing.h]: updated

File: aten/src/ATen/TensorIndexing.h (hunk 3 above)
Comment (tugsbayasgalan) [reply aten/src/ATen/TensorIndexing.h]: updated

File: aten/src/ATen/TensorIndexing.h (hunk 4 above)
Comment (tugsbayasgalan) [reply aten/src/ATen/TensorIndexing.h]: updated

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42049013).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42049013).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42049013).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42049013).

**Comment by ezyang** [comment]
This is obsoleted by https://github.com/pytorch/pytorch/pull/91341 right?

**Pull request closed without merging**